File: src/main/scala/com/keithalcock/torchscript/NerVocab.scala
package com.keithalcock.torchscript

import com.keithalcock.torchscript.utils.Closer.AutoCloser
import com.keithalcock.torchscript.utils.StringUtils

import scala.io.Source

object NerVocab {
  val mask = "<MASK>"

  def apply(embeddingPath: String): Map[String, Int] = {
    val words = Source.fromFile(embeddingPath).autoClose { source =>
      source.getLines.drop(1).map { line =>
        StringUtils.beforeFirst(line, ' ')
      }.toVector // must be realized before close
    }
    // This first one isn't technically sorted.
    val sortedWords = mask +: words.sorted
    // Add one so that we can preserve index 0 for our padding
    val vocab = sortedWords.zipWithIndex.map { case (word, index) =>
      // println(s"${index}\t$word")
      word -> index
    }.toMap

    vocab
  }
}

File: src/main/python/NerVocab.py
mask = "<MASK>"

def read_vocab(embedding_path):
    vocab = {}

    # Load word embeddings, (we just care about which tokens are in our vocab)
    with open(embedding_path, "r") as embed_file:
        lineno = 0
        for line in embed_file:
            if lineno > 0:
                cur = line[:line.index(" ")]
                vocab[cur] = -1
            lineno += 1

    # Add one so that we can preserve index 0 for our padding
    for i, token in enumerate(sorted(vocab.keys())):
        vocab[token] = i + 1
        # print(f"{i + 1}\t{token}")
    vocab[mask] = 0
    # print(f"{vocab[mask]}\t{mask}")

    return vocab
